docs/2023/puzzles/day04.md

```diff
@@ -2,10 +2,139 @@ import Solver from "../../../../../website/src/components/Solver.js"
+  val givenNumbers = givenNumberStrs.drop(1).map(_.toInt).toSet
+```
+
+Although this is not specified in the puzzle description, it seems like a number
+can appear at most once in the list of winning numbers as well as in the given
+numbers, so it is valid to perform `toSet` and the final counting step, which is
+what we return:
+
+```scala
+  winningNumbers.intersect(givenNumbers).size
+end countWinning
+```
+
+Lastly, the standard library conveniently gives us an iterator over lines.
+
+```scala
+def winningCounts(input: String): Iterator[Int] =
+  input.linesIterator.map(countWinning)
+```
+
+### Part 1
+
+```scala
+def part1(input: String): String =
+  winningCounts(input)
+```
+
+The next line is the core logic for this part.
+
+```scala
+    .map(winning => if winning > 0 then Math.pow(2, winning - 1).toInt else 0)
+    .sum.toString()
+end part1
+```
+
+### Part 2
+
+```scala
+def part2(input: String): String =
+  winningCounts(input)
+```
+
+The key insight here is that when we process one card, the cards we win (if any)
+are always at the immediately following indices. So instead of keeping track of
+*absolute* indices (e.g. "3 copies of card 5"), we only keep track of how many
+cards we've won *relative to the current index* (e.g. "3 copies of the
+next-to-next card"). This is the `Vector` in the accumulator of our fold.
+
+```scala
+    .foldLeft((0, Vector(1))){ case ((numCards, multiplicities), winning) =>
+```
+
+In each iteration, we remove its first element, i.e. the multiplicity of the
+current card...
+
+```scala
+      val thisMult = multiplicities(0)
+```
+
+... and carry forward the rest:
+
+```scala
+      val restMult = multiplicities
+        .drop(1)
+```
+
+If we just won no new cards, then we extend the vector by a single `1` for the 1
+original copy of the next card to be processed in the next iteration. Else, we
+extend the vector by as many elements as required to keep track of the cards we
+just won.
+
+```scala
+        .padTo(Math.max(1, winning), 1)
+```
+
+Remember that we win a copy of a later card for every copy we'd already won of
+the current card.
+
+```scala
+        .zipWithIndex
+        .map((mult, idx) => if idx < winning then mult + thisMult else mult)
+      (numCards + thisMult, restMult)
+    }
+    ._1.toString()
+end part2
+```
+
+Throughout the iteration, the vector satisfies the following invariants:
+* It has at least one element.
+* All its elements are positive.
+* When processing a card, the first element is the final multiplicity of that
+  card. (It is summed up in `numCards` in the accumulator.)
+
+Why track by relative index instead of absolute?
+* We don't have to parse or store the card numbers.
+* We can discard information as soon as it is no longer needed, and keep only
+  limited information about the future, in this case bounded by the maximum
+  possible number of winning numbers.
+* (Personal opinion) It makes for a nicer, purely functional solution!
+
```

Comment:
Its good also to include the code in one whole chunk here:
~~~markdown
## Final Code

```scala
val foo = ???
```
~~~